lib/bech32/test/Codec/Binary/Bech32Spec.hs

```diff
@@ -374,17 +387,25 @@ invalidChecksums =
       , Bech32.StringToDecodeContainsInvalidChars [CharPosition 41] )
     ]
 
-newtype ValidBech32Char = ValidBech32Char
-    { getValidBech32Char :: Char
+newtype DataChar = DataChar
+    { getDataChar :: Char
     } deriving (Eq, Ord, Show)
 
-instance Arbitrary ValidBech32Char where
-    arbitrary = ValidBech32Char <$> elements Bech32.charset
-    shrink (ValidBech32Char c) =
-        ValidBech32Char . (Bech32.word5ToChar Arr.!) <$> shrink
+instance Arbitrary DataChar where
+    arbitrary = DataChar <$> elements Bech32.dataCharList
+    shrink (DataChar c) =
+        DataChar . Bech32.dataCharFromWord <$> shrink
```

Comment:
This shrinker seems a bit "clunky" now that I see it. We shrink `Word5` by shrinking `Word8` and calling `Bech32.word5` on them. Internally, `word5` does a bitwise and `31` to nullify the first three bits. which means that in practice, there's a whole range of `Word8` that gets shrinked to the same `Word5`. And therefore, I believe it's quite easy to end up shrinking to the _same_ `Word5` by just accidentally picking in wrong range of `Word8`. 

This would cause QC to loop endlessly trying to shrink a `DataChar` whereas there's very little value here in shrinking such element. What about just yielding an empty list?

Reply:
> And therefore, I believe it's quite easy to end up shrinking to the same `Word5` by just accidentally picking in wrong range of `Word8`.

I've added a new commit to fix this. Instead of masking arbitrary `Word8` values to generate our values of `Word5` (which would give a skewed distribution), I instead use `arbitraryBoundedEnum`, and also provide `Enum` and `Bounded` instances for `Word5`.

The `arbitraryBoundedEnum` function yields a uniform distribution `[minBound .. maxBound]`, which in the case of `Word5`, is `[0 .. 31]`. 

The shrinker should now work, as internally it just calls [`shrinkIntegral`](http://hackage.haskell.org/package/QuickCheck-2.13.1/docs/src/Test.QuickCheck.Arbitrary.html#shrinkIntegral), which will always yield values in the range `[0 .. 30]`. Such values will always fit within the range of a `Word5`.